cpp/src/join/mixed_join_kernels.cuh

```diff
@@ -0,0 +1,319 @@
+ * the kernel needs to internally loop over left rows. Otherwise, loop over right rows.
+ * @param[out] output_size The resulting output size
+ */
+template <int block_size, bool has_nulls>
+__global__ void compute_mixed_join_output_size(
+  table_device_view left_table,
+  table_device_view right_table,
+  table_device_view probe,
+  table_device_view build,
+  null_equality compare_nulls,
+  join_kind join_type,
+  cudf::detail::mixed_multimap_type::device_view hash_table_view,
+  ast::detail::expression_device_view device_expression_data,
+  bool const swap_tables,
+  std::size_t* output_size,
+  cudf::size_type* matches_per_row)
+{
+  // The (required) extern storage of the shared memory array leads to
+  // conflicting declarations between different templates. The easiest
+  // workaround is to declare an arbitrary (here char) array type then cast it
+  // after the fact to the appropriate type.
+  extern __shared__ char raw_intermediate_storage[];
+  cudf::ast::detail::IntermediateDataType<has_nulls>* intermediate_storage =
+    reinterpret_cast<cudf::ast::detail::IntermediateDataType<has_nulls>*>(raw_intermediate_storage);
+  auto thread_intermediate_storage =
+    intermediate_storage + (threadIdx.x * device_expression_data.num_intermediates);
+
+  std::size_t thread_counter{0};
+  cudf::size_type const start_idx      = threadIdx.x + blockIdx.x * blockDim.x;
+  cudf::size_type const stride         = blockDim.x * gridDim.x;
+  cudf::size_type const left_num_rows  = left_table.num_rows();
+  cudf::size_type const right_num_rows = right_table.num_rows();
+  auto const outer_num_rows            = (swap_tables ? right_num_rows : left_num_rows);
+
+  auto evaluator = cudf::ast::detail::expression_evaluator<has_nulls>(
+    left_table, right_table, device_expression_data);
+
+  // TODO: The hash join code assumes that nulls exist here, so I'm doing the
+  // same but at some point we may want to benchmark that.
+  row_hash hash_probe{nullate::YES{}, probe};
+  auto const empty_key_sentinel = hash_table_view.get_empty_key_sentinel();
+  make_pair_function pair_func{hash_probe, empty_key_sentinel};
+
+  for (cudf::size_type outer_row_index = start_idx; outer_row_index < outer_num_rows;
+       outer_row_index += stride) {
+    // Figure out the number of elements for this key.
+    cg::thread_block_tile<1> this_thread = cg::this_thread();
+    auto query_pair                      = pair_func(outer_row_index);
+    // TODO: Address asymmetry in operator.
+    auto count_equality = pair_expression_equality<has_nulls>{
+      build, probe, evaluator, thread_intermediate_storage, compare_nulls};
+    // TODO: This entire kernel probably won't work for left anti joins since I
+    // need to use a normal map (not a multimap), so this condition is probably
+    // overspecified at the moment.
+    if (join_type == join_kind::LEFT_JOIN || join_type == join_kind::LEFT_ANTI_JOIN ||
+        join_type == join_kind::FULL_JOIN) {
```

Comment:
`join_type` is used to determine whether it's an outer join or not. Similarly to hash joins, what do you think if we move this logic (runtime comparison) out of the kernel and pass an `is_outer` via template instead? The downside is that it will be more painful to compile, but this can get rid of runtime comparison and if statements.

Reply:
I think it's probably worth testing your suggestions for templating on both `join_type` and `is_outer` just to see what the compile/run time tradeoffs are. I suspect that the performance improvements will be minimal relative to the compile-time blowup, but it's worth trying and seeing. It would really be nice for us to prioritize speeding up cuco multimap compilation though, I think the multimap is particularly slow, significantly more so than the static_map based on both my experience incorporating them into cudf and the development on cuco itself.

Reply:
> the performance improvements will be minimal relative to the compile-time blowup

Agreed


> It would really be nice for us to prioritize speeding up cuco multimap compilation

That's my plan during this Christmas holiday.

Reply:
In the interest of getting a working version of this feature out ASAP (and to focus on the corresponding feature for semi/anti joins), and given that we'll probably go through multiple rounds of future refactoring of cuco and these kernels (especially when we work on #9557) I'm going to postpone testing these compile-time improvements to a future PR.